src/TopPage/DataExtension.php

```diff
@@ -238,4 +298,88 @@ protected function clearTopPage(): void
     {
         $this->owner->TopPageID = 0;
     }
+
+    /**
+     * Assigns top page relation based on fixed id
+     *
+     * @see DataExtension::withFixedTopPage()
+     */
+    protected function assignFixedTopPage(): void
+    {
+        $this->owner->TopPageID = $this->getFixedId();
+    }
+
+    /**
+     * Save top page changes without using write()
+     * Using raw query here because:
+     * - this is already called during write() and triggering more write() related extension points is undesirable
+     * - we don't want to create a new version if object is versioned
+     * - using writeWithoutVersion() produces some weird edge cases were data is not written
+     * because the fields are not recognised as changed (using forceChange() introduces a new set of issues)
+     */
+    protected function saveChanges(): void
+    {
+        /** @var DataObject|DataExtension $owner */
+        $owner = $this->owner;
+
+        $table = $this->getTopPageTable();
+
+        if (!$table) {
+            return;
+        }
+
+        $query = SQLUpdate::create(
+            sprintf('"%s"', $table),
+            [
+                '"TopPageID"' => $owner->TopPageID,
+            ],
+            ['"ID"' => $owner->ID]
+        );
+
+        $query->execute();
+    }
+
+    /**
+     * Perform a page lookup based on cached data
+     *
+     * @param int $id
+     * @return Page|null
+     */
+    protected function getTopPageFromCachedData(int $id): ?Page
```

Comment:
This function can be deleted, just call `$page = Page::get_by_id($id);` in the one place this is used

Reply:
This is intentionally here to allow extensibility. Bespoke projects may override this method to customise how it works. I was thinking about adding an extension point here but it's not flexible enough as you may want to completely alter the behaviour here as opposed to just add stuff on top of the default. This is a performance driven feature after all. Example cases:

* inject tracking of this feature (tracks when this feature looks up pages from cached data)
* feature flag / enable / disable this feature
* load different page / page from different data source

I'll update the docs here to make it more clear why this is needed.